client/src/three/components/StructureManager.ts

```diff
@@ -88,18 +120,73 @@ export class StructureManager {
     const index = this.structures.addStructure(entityId, key);
 
     if (this.structureModels) {
-      const modelType = this.structureModels.get(key);
-      modelType?.setMatrixAt(index, this.dummy.matrix);
-      modelType?.setCount(this.structures.getCountForType(key)); // Set the count to the current number of structures
-
-      // Add label on top of the structure with appropriate color
-      const labelColor = isMine ? myColor : neutralColor;
-      const label = this.labelManagers.get(key)?.createLabel(position as any, labelColor);
-      this.scene.add(label!);
+      const models = this.structureModels.get(key);
+      if (models && models.length > 0) {
+        const modelIndex = this.getModelIndex(structureType, entityId);
+        const modelType = models[modelIndex];
+
+        modelType.setMatrixAt(index, this.dummy.matrix);
+        modelType.setCount(this.structures.getCountForType(key));
+
+        // Add label on top of the structure with appropriate color
+        const labelColor = isMine ? myColor : neutralColor;
+        const label = this.labelManagers.get(key)?.createLabel(position as any, labelColor);
+        this.scene.add(label!);
+      }
     }
   }
+
+  getModelIndex(structureType: StructureType, entityId: ID): number {
```

Comment:
getModelIndex is just for Hyperstructure progress? if so name should be more explicit